test/k4FWCoreTest/src/components/ExampleFunctionalConsumer.cpp

```diff
@@ -0,0 +1,42 @@
+#include "Gaudi/Algorithm.h"
+#include "Gaudi/Property.h"
+#include "GaudiAlg/Consumer.h"
+#include "k4FWCore/DataWrapper.h"
+
+#include "edm4hep/MCParticleCollection.h"
+#include "podio/CollectionBase.h"
+
+#include <string>
+
+// This will always be Gaudi::Algorithm
+using BaseClass_t = Gaudi::Functional::Traits::BaseClass_t<Gaudi::Algorithm>;
+
+// Which type of collection we are reading
+// this will always be podio::CollectionBase
+// Has to be wrapped in DataWrapper
+using colltype = DataWrapper<podio::CollectionBase>;
```

Comment:
As just discussed in person you can avoid these using statements and the user side visibility of handles and wrappers by changing BaseClass_t.
We are using our special wrappers because of what is needed for PODIO [here](https://github.com/key4hep/k4FWCore/blob/338f4ec16d29a8e0d2b2ae98ea781d06fa62d7a8/k4FWCore/src/PodioDataSvc.cpp#L125). 
To make Gaudi Functional use these, we have to make them visible to `[Consumer/Producer/Transformer]`. which internally use `DataHandleMixin`. The latter uses this [template magic](https://gitlab.cern.ch/gaudi/Gaudi/-/blob/master/GaudiFunctional/include/Gaudi/Functional/details.h#L419) to know which Handles to use. To override the default we need to subclass of `Gaudi::Algorithm` that adds something like 
 
```
template<typename T>using InputHandle_t = DataObjectReadHandle<DataWrapper<T>>;
```
and a similar one for the writing part. 

Doing this we will not be able to use the legacy read/write any more though. If we want to preserve that, we have to use the key4hep specific data handle which we wanted to deprecate.

Reply:
I was able to do this for the producers. When trying to do the same thing for the consumers (or transformers since they also have inputs) I get the following error message:
```
ExampleFunction...  ERROR Wrong DataObjectType : The type expected for /Event/MCParticles is AnyDataWrapper<edm4hep::MCParticleCollection> and is different from the one of the object in the store which is DataWrapper<podio::CollectionBase>.
```
that is coming from Gaudi performing a check with a `dynamic_cast`: https://gitlab.cern.ch/gaudi/Gaudi/-/blob/master/GaudiKernel/include/GaudiKernel/DataObjectHandle.h#L56. Options:
- Templated registering of objects, not trivial to do
- Using a wrapper around `Gaudi::Functional::Consumer` (and transformer) that retrieves the collection as `podio::CollectionBase` and then casts to the right type, but I like that our algorithms and the ones from Gaudi look the same
- Maybe it's also possible to have a wrapper that we can use for registering and then wrappers of concrete types can inherit from so that the `dynamic_cast` check would pass.

Ideally I would like it to be like it is in the example producers right now, where we inherit from the corresponding `Gaudi::Functional` and we just have to specify the type of collection that we want.

Reply:
> Ideally I would like it to be like it is in the example producers right now, where we inherit from the corresponding Gaudi::Functional and we just have to specify the type of collection that we want.

I agree with this. In case it is not possible the option I would like the second most after this one is to have a `k4FWCore::Functional` (naming up for discussion) that is effectively a template specialization of the corresponding `Gaudi::Functional` that injects the necessary types for us. I don't know if this is something that is possible, or if would always need a wrapper.

Reply:
The `maybeRead` is there to simplify the interface and not have to use `podio::CollectionBase` as type and then do a dynamic cast by pushing directly, for example, an `edm4ep::MCParticleCollection` to the TES. There is maybe an option to create a custom Functional base class we can inherit from in k4FWCore that somehow gets around this so that it is possible to retrieve a `podio::CollectionBase` and do the relevant dynamic casting inside the class and not have the `maybeRead` stuff, but I don't think this is trivial to do